common/paramsbuilder/module.go

```diff
@@ -11,52 +13,57 @@ var ErrNoSupportedModule = errors.New("no supported module was chosen")
 // This is relevant where there are several APIs for different sub-products, and the APIs
 // are versioned differently or have different ways of constructing URLs and requests for reading/writing.
 type Module struct {
-	// Name represents the name of the sub-product. e.g. crm, marketing, etc.
-	Name string
+	// ID is name of the module.
+	ID APIModuleID
+	// Path represents the URL path of the sub-product. e.g. crm, marketing, etc.
+	Path string
 	// Connector implementation must provide list of modules that are currently supported.
 	// This defines a list of module a user could switch to. Any out of scope module will be ignored.
-	supported []APIModule
+	supported APIModuleMapping
 	// If user supplied unsupported module it will use this fallback.
-	fallback *APIModule
+	fallbackID APIModuleID
 }
 
 func (p *Module) ValidateParams() error {
 	// making sure the provided module is supported.
 	// If the provided module is not supported, use fallback.
-	if !p.isSupported() {
-		if p.fallback == nil {
-			// not supported and user didn't provide a fallback
-			return ErrNoSupportedModule
-		}
+	if apiModule, ok := p.supported[p.ID]; ok {
+		p.save(p.ID, apiModule)
+
+		return nil
+	}
 
+	if fallbackModule, ok := p.supported[p.fallbackID]; ok {
 		// replace with fallback module
-		p.Name = p.fallback.String()
+		p.save(p.fallbackID, fallbackModule)
 
-		// even fallback is not supported
-		if !p.isSupported() {
-			return ErrNoSupportedModule
-		}
+		return nil
 	}
 
-	return nil
+	// Neither selected module nor the fallback module are supported.
+	// It is expected that fallback was correctly included by connector developer.
+	return ErrNoSupportedModule
 }
 
-func (p *Module) WithModule(module APIModule, supported []APIModule, defaultModule *APIModule) {
-	p.Name = module.String()
+// WithModule allows API module selection.
+// Connector implementation must provide list of modules that are currently supported.
+// This defines a list of module a user could switch to. Any out of scope module will be ignored.
+// If user supplied unsupported module, then it will use this fallback.
+func (p *Module) WithModule(moduleID APIModuleID, supported APIModuleMapping, fallbackID APIModuleID) {
+	p.ID = moduleID
 	p.supported = supported
-	p.fallback = defaultModule
+	p.fallbackID = fallbackID
 }
 
-func (p *Module) isSupported() bool {
-	for _, mod := range p.supported {
-		if p.Name == mod.String() {
-			return true
-		}
-	}
-
-	return false
+func (p *Module) save(id APIModuleID, m APIModule) {
+	p.ID = id
+	p.Path = m.String()
 }
 
+type APIModuleID string
+
+type APIModuleMapping handy.Map[APIModuleID, APIModule]
```

Comment:
The mapping is an implementation detail, but for actual users, it would just look like listing out all the possible modules which is why I feel like just `Modules` is a better type name